test-data/unit/check-python38.test

```diff
@@ -362,6 +362,36 @@ reveal_type(z2)  # E: Name 'z2' is not defined  # N: Revealed type is 'Any'
 
 [builtins fixtures/isinstancelist.pyi]
 
+[case testWalrusConditionalTypeBinder]
+from typing import Union
+from typing_extensions import Literal
+
+class Good:
+    @property
+    def is_good(self) -> Literal[True]: ...
+
+class Bad:
+    @property
+    def is_good(self) -> Literal[False]: ...
+
+def get_thing() -> Union[Good, Bad]: ...
+
+if (thing := get_thing()).is_good:
+    reveal_type(thing)  # N: Revealed type is '__main__.Good'
+else:
+    reveal_type(thing)  # N: Revealed type is '__main__.Bad'
+[builtins fixtures/property.pyi]
+
+[case testWalrusConditionalTypeCheck]
+from typing import Optional
+
+maybe_str: Optional[str]
+
+if (is_str := maybe_str is not None):
```

Comment:
Does this PR affect simple cases, such as `if x := f()`? If yes, it's worth testing these separately.

What about `if (x := f()) is not None` (note that parentheses are different from this test case)?

Reply:
I added tests for both `if x := f()` and `if (x := f()) is not None` cases.

Also, I found two odd behaviors:

- Type narrowing without `--strict-optional` flags seems odd because it can narrows the `Union` outside of conditional statement even though it is not known which value the variable actually holds. With `--strict-optional` the last `reveal_type` in below example emits `Union[builtins.str, None]`.

```python
from typing import Optional

maybe_str: Optional[str]

if (x := maybe_str is not None):
    reveal_type(maybe_str)  # N: Revealed type is 'builtins.str'
else:
    reveal_type(maybe_str)  # N: Revealed type is 'None'

reveal_type(maybe_str)  # N: Revealed type is 'builtins.str'
```

- Below case is not handled properly and does not reflect the runtime behavior. Note that this is actually a different use case than the one I added in `testWalrusAssignmentAndConditionScopeForLiteral` test case because it does not fall into `is_true_literal` or `is_false_literal` checks in `find_isinstance_check_helper` method.

```python
if (x := 5):
    reveal_type(x)  # N: Revealed type is 'builtins.int'
else:
    reveal_type(x)  # E: Statement is unreachable

reveal_type(x)  # N: Revealed type is 'builtins.int'
```

Reply:
> if (x := maybe_str is not None):

Note that this is parsed as “x := (maybe_str is not None)“